File: 2016/Day 02/day02.cpp
//
//  Advent of Code 2016
//  Day 01
//
//  By PatchesTheDipstick
//


//  System includes...

#include <vector>
#include <map>
#include <utility>
#include <string>

//  Third Party includes...



//  Project includes...

#include "../utils.hpp"

int newButton(int current_button, std::string direction)
{
	int return_value = current_button;
	
	if (direction == "U")
	{
		if (return_value > 3)
		{
			return_value -= 3;
		}
	}
	else if (direction == "L")
	{
		if ((return_value != 1) && (return_value != 4) && (return_value != 7))
		{
			return_value--;
		}
	}
	else if (direction == "D")
	{
		if (return_value < 7)
		{
			return_value += 3;
		}
	}
	else if (direction == "R")
	{
		if ((return_value != 3) && (return_value != 6) && (return_value != 9))
		{
			return_value++;
		}
	}
	//std::cout << return_value << std::endl;
	return return_value;
}

int newButtonTwo(int current_button, std::string direction)
{
	int return_value = current_button;
	if (direction == "U")
	{
		if (return_value == 3)
		{
			return_value = 1;
		}
		else if ((return_value >= 6) && (return_value <= 8))
		{
			return_value -= 4;
		}
		else if ((return_value >= 10) && (return_value <= 12))
		{
			return_value -= 4;
		}
		else if (return_value == 13)
		{
			return_value -=2;
		}
	}
	else if (direction == "L")
	{
		if ((return_value >= 3) && (return_value <= 4))
		{
			return_value -=1;
		}
		else if ((return_value >= 6) && (return_value <= 9))
		{
			return_value -=1;
		}
		else if ((return_value >= 11) && (return_value <= 12))
		{
			return_value -= 1;
		}
	}
	else if (direction == "D")
	{
		if (return_value == 1)
		{
			return_value = 3;
		}
		else if ((return_value >= 2) && (return_value <= 4))
		{
			return_value += 4;
		}
		else if ((return_value >= 6) && (return_value <= 8))
		{
			return_value += 4;
		}
		else if (return_value == 11)
		{
			return_value = 13;
		}
	}
	else if (direction == "R")
	{
		if ((return_value >= 2) && (return_value <= 3))
		{
			return_value +=1;
		}
		else if ((return_value >= 5) && (return_value <= 8))
		{
			return_value +=1;
		}
		else if ((return_value >= 10) && (return_value <= 11))
		{
			return_value += 1;
		}
	}
	return return_value;
}

int main(int argc, char** argv)
{
	std::vector<std::string> input = ptd::utils::getInput("input.txt");
	std::vector<int> bathroom_code;
	std::vector<int> part_two_code;
	int current_button = 5;
	int part_two_button = 5;
	for (int a = 1; a < input.size(); a++)
	{
		for (int b = 0; b < input[a].size(); b++)
		{
			current_button = newButton(current_button, input[a].substr(b, 1));
			part_two_button = newButtonTwo(part_two_button, input[a].substr(b, 1));
		}
		bathroom_code.push_back(current_button);
		part_two_code.push_back(part_two_button);
	}
	
	
	std::cout << "Advent of Code Day 02:" << std::endl;
	std::cout << "    Part 1 : ";
	for (int a = 0; a < bathroom_code.size(); a++)
	{
		std::cout << bathroom_code[a];
	}
	std::cout << std::endl;
	std::cout << "    Part 2 : ";
	for (int a = 0; a < part_two_code.size(); a++)
	{
		std::cout <<  std::hex << part_two_code[a];
	}
	std::cout << std::endl;

	return (0);

}

File: 2016/Day 02/day02.py
#
#  Advent of Code 2016
#  Day 02
#
#  By PatchesTheDipstick
#

def getInput(file_name):
	input = []
	with open(file_name) as input_file:
		input = input_file.read().splitlines()
	return input

def part_one(directions):
	keypad = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
	code = ""
	x = y = 1
	for dir in directions:
		for d in dir:
			if d == 'U':
				if y > 0:
					y -=1
			elif d == 'D':
				if y < 2:
					y += 1
			elif d == 'L':
				if x > 0:
					x -= 1
			elif d == 'R':
				if x < 2:
					x += 1
		code += str(keypad[y][x])
	return code

def part_two(directions):
	keypad = []
	keypad.append(['N', 'N', '1', 'N', 'N'])
	keypad.append(['N', '2', '3', '4', 'N'])
	keypad.append(['5', '6', '7', '8', '9'])
	keypad.append(['N', 'A', 'B', 'C', 'N'])
	keypad.append(['N', 'N', 'D', 'N', 'N'])
	r = 2
	c = 0
	code = ""
	for dir in directions:
		for d in dir:
			if d =='U':
				if c in [1, 3] and r > 1:
					r -=1
				elif c == 2 and r > 0:
					r -= 1
			elif d == 'D':
				if c in [1, 3] and r < 3:
					r +=1
				elif c == 2 and r < 4:
					r += 1
			elif d =='L':
				if r in [1, 3] and c > 1:
					c -=1
				elif r == 2 and c > 0:
					c -= 1
			elif d == 'R':
				if r in [1, 3] and c < 3:
					c +=1
				elif r == 2 and c < 4:
					c += 1
		code += keypad[r][c]
	return code


if __name__ == '__main__':
	
	input = getInput("input.txt")
	
	print("Advent of Code 2016 Day 02")
	print("  Part one:", part_one(input))
	print("  Part two:", part_two(input))	
	